ML / VLLM / inference related configs

KRRT7 opened the pull request:
# Description

## ❓ What does this PR do?

Add anti-bloat plugin configurations to strip IPython dependencies from Roboflow modules when IPython support is disabled.

The added configurations cover:

- `roboflow.core.version`, removing IPython references when IPython support is not used
- `roboflow.models.object_detection`, replacing `ipywidgets` imports and ignoring IPython when IPython support is not used

## 🧐 Why was it initiated? Any relevant Issues?

This was initiated to improve ML, VLLM, and inference-related package handling by avoiding unnecessary IPython dependencies in Roboflow configurations.

## 🧱 Type of Change

- [ ] 🐛 Bug fix (non-breaking change which fixes an issue)
- [x] ✨ New feature (non-breaking change which adds functionality)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] 🧹 Refactoring (no functional changes, no api changes)
- [ ] 🏗️ Build / CI System
- [ ] 📚 Documentation Update

## ✅ PR Checklist

- [x] **Correct base branch selected**: Should be `develop` branch.
- [ ] **Formatting**: Enabled commit hook or executed `./bin/autoformat-nuitka-source`.
- [ ] **Tests**: All tests still pass. (See [Running the Tests](https://nuitka.net/doc/developer-manual.html#running-the-tests)).
  - Existing PR body did not list local validation commands.
- [ ] **New Coverage**: New features or fixed regressions are covered via new tests.
- [ ] **Documentation**: Documentation updates are included for new or changed features.

## 🤖 AI Generated Code Policy

- [ ] **Detection**: This PR contains AI generated code.
  - [ ] **Issue First**: I have created an issue explaining the problem *before* using AI to generate a fix, ensuring the direction is correct.
  - [ ] **Documentation**: I have provided the prompts used to generate the code (non-optional).
  - [ ] **Verification**: I have **manually verified** the AI generated code.
  - [ ] **Evidence**: I have included test evidence that the change is effective.

## Summary by Sourcery

Add anti-bloat plugin configurations to strip IPython dependencies from Roboflow modules when IPython support is disabled.

Enhancements:
- Introduce anti-bloat rules for roboflow.core.version to remove IPython references when not using IPython.
- Add anti-bloat rules for roboflow.models.object_detection to replace ipywidgets imports and ignore IPython when not using IPython.

KRRT7 committed:
add roboflow config

kayhayen commented on nuitka/plugins/standard/standard.nuitka-package.config.yml:
```diff
@@ -9482,3 +9482,19 @@
+  anti-bloat:
+    - description: 'remove IPython reference'
+      replacements_plain:
+        'import ipywidgets.widgets as widgets': 'raise ImportError'
```
I added no-auto-follow for exactly this purpose with the added benefit that it's a global config, i.e. you just attach it to a package. If a package is optional for one submodule, very generally it is for all uses and when things move, we then don't care.

As a result it's more simple:

```
- module-name: 'roboflow'
  anti-bloat:
    - description: 'remove IPython reference'
      no-auto-follow:
        'IPython': 'ignore'
        'ipywidgets': 'ignore'
      when: 'not use_ipython'
```

kayhayen committed:
New stable release.

kayhayen committed:
Merge branch 'release/2.8'

kayhayen committed:
Merge tag '2.8' into develop

kayhayen committed:
New release cycle.

kayhayen committed:
Standalone: Honor "--nofollow-import-to" for stdlib modules as well

* This allows users to manually reduce the standard library usage too

kayhayen committed:
Anti-Bloat: Don't allow custom "nofollow" modes, point the user to the correct option instead

kayhayen committed:
Onefile: Fix, need to still register signal handlers even in DLL mode

kayhayen committed:
Onefile: Allow to disable onefile timeout and hard killing on CTRL-C entirely

kayhayen committed:
Release: Remove Python2 packages from CI container

kayhayen committed:
fixup! Accelerated: Enhanced detection for uninstalled Python for Anaconda and WinPython

kayhayen committed:
Debugging: Adding rusage capture for OSes that support it

* Only POSIX compliant OSes will do it, Linux, macOS, *BSD, but not Android.

* Not yet part of the report, need to figure out how to use the information, that also won't be diffable at all.

kayhayen committed:
Windows: Fix, need to convert paths with forward slashes for cache directories

* This was making Nuitka-Action fail over.

kayhayen committed:
Anti-Bloat: Avoid warning during program shutdown of compiled "xgboost" package

kayhayen committed:
New hotfix release.

kayhayen committed:
Debugging: In case of tool commands failing, output the too long command line

kayhayen committed:
Merge branch 'hotfix/2.8.1'

kayhayen committed:
Merge tag '2.8.1' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
fixup! New release cycle.

kayhayen committed:
Optimization: Find previous assignment traces faster

* The assignment and del nodes were using function to find what they
  already knew.

* The "self.variable_trace" already kept track of the previous one,
  and in case of overwriting just needs to be preserved.

* For matching unescaped traces we will do similar, but it's not
  really used right now, so make it only a TODO.

* This speeds up the first micro pass even more, because it doesn't
  have to search and do other things, if not previous trace exists.

* Also the common check if no by name or merges of a value occurred
  was used inverted and now should be slightly faster to use.

* While this accelerated the first micro pass by a lot for per
  assignment work, it mainly means to cleanup the design such
  that traces are easier to re-recognize. And this is a first
  step with immediate impacts.

KRRT7 committed:
Scons: Added experimenntal support for enabling zhin flto with Clang

* Using an experimental option.

* This is to make it possible to evaluate if this makes sense as a new default once we have
  rusage data in compilation reports to tell from there.
---------

Co-authored-by: Kay Hayen <kay.hayen@gmail.com>

kayhayen committed:
Onefile: Fix, insist on output path for onefile binary to exist with "--output-dir" option

* It was insisting on the wrong folder, and then crashing because it
  later cannot be converted or created, at least on Windows.

kayhayen committed:
macOS: Fix, could crash when a lot of package specific directories were used

* Since we now scan for all of Homebrew sub-folders this could become
  a lot directories, resulting in too many arguments error when calling
  "otool".

* The environment need not be set for the call though, which avoids
  that issue and makes this also a cleanup.

kayhayen committed:
fixup! Reports: Include environment flags for C compiler and linker picked up

kayhayen committed:
Minor cleanup

kayhayen committed:
UI: Was using wrong logger for message

kayhayen committed:
Plugins: Need to consider "global-sys-path" contributions for compile time code

* When checking package versions of vendored code, these wouldn't be found.

* Also code context for anti-bloat replacements wasn't able to import these packages.

* This is necessary for at least the "ray" package and its use of "psutil", but probably affects other cases too.

kayhayen committed:
Standalone: Added support for "ray" package (WIP)

* Their vendored packages need to be configured and there are hidden
  dependencies for "ray._raylet".

kayhayen committed:
Standalone: Added implicit dependency for "oracledb"

kayhayen committed:
New hotfix release.

kayhayen committed:
UI: Actually --output-filename works fine for standalone mode too

kayhayen committed:
Scons: Make zig work as a C11 compiler (WIP)

* With zig 15.2 at least the atomics now work fine, which removes
  important barriers.

* Need to disable a few warnings and tweaks to be warning free.

* Also does C23 file embedding so use that too.

* The way the "cc" is injected depends on "ccache" being used, and
  we need to clean that up later, so Windows works too.

kayhayen committed:
UI: Fix, accelerated mode didn't honor "--output-dir" if "--output-filename" was given as well

kayhayen committed:
Minor spelling cleanups

kayhayen committed:
Scons: Make zig work for Windows too (WIP)

* Need to consider if using mingw, clang, or zig more carefully.

* With this standalone and onefile mode are working.

* Also added control over zig cache file locations to be below
  Nuitka's cache directory.

kayhayen committed:
Merge branch 'hotfix/2.8.2'

kayhayen committed:
Merge tag '2.8.2' into develop

kayhayen committed:
fixup! Onefile: Fix, insist on output path for onefile binary to exist with "--output-dir" option

kayhayen committed:
New hotfix release.

kayhayen committed:
Merge branch 'hotfix/2.8.3'

kayhayen committed:
Merge tag '2.8.3' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
Python Build Standalone: Added built-in modules for debug build to not raise

* This adds another one observed with 3.11 at least.

kayhayen committed:
Debugging: Added assertions for code object creation

* We are getting assertions from Python when built with zig,
  and these are supposed to do those as well.

kayhayen committed:
macOS: Added support for newer PySide6

kayhayen committed:
Scons: Add support for using zig on macOS too (WIP)

kayhayen committed:
New hotfix release.

kayhayen committed:
Merge branch 'hotfix/2.8.4'

kayhayen committed:
Merge tag '2.8.4' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
macOS: Fix, make sure to use system "xattr" binary

* Otherwise using "arch -x86_64 python" for compilation
  can fail when some packages are installed that provide
  "xattr" too, as that is a "arm64" binary only.

kayhayen committed:
macOS: Fix, avoid setting $ORIGIN rpaths that end up unused

* Also solve partially a TODO of minimizing intermediate
  directories in rpaths of ELF platforms, by only putting
  them there is they contain DLLs. This reduces them in
  many cases entirely.

kayhayen committed:
UI: Don't say "--include-data-files-external" doesn't work in standalone mode

* It does for a while, but the help still said it wouldn't.

kayhayen committed:
Standalone: Only include metadata from Nuitka Package Configuration if the package is included

* Otherwise at runtime crashes can occur when creating entry point objects.

* User provided inclusions are not subject to this filtering.

kayhayen committed:
New hotfix release.

kayhayen committed:
Scons: Detect zig version

* Also call the actually used Clang, not the one in PATH for version output

kayhayen committed:
Fix, do not use zero length dynamic local arrays with zig compiler

* That's the only one that doesn't support this non-standard extension.

kayhayen committed:
Accelerated: Also honor "--output-filename" and" --output-dir" for created runner scripts

* The ".cmd" or ".sh" file was not having the correct name nor#
  was it created in the directory specified.

kayhayen committed:
fixup! Standalone: Add support for including vc runtime dlls on Windows

kayhayen committed:
Standalone: Added more "metadata" inclusion for "transformers" package.

kayhayen committed:
UI: Fix typo in 3.13.4 rejection message

kayhayen committed:
Plugins: Detect plugins that are detected by namespace usage for module mode as well

kayhayen committed:
Merge branch 'hotfix/2.8.5'

kayhayen committed:
Merge tag '2.8.5' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
Standalone: Fix, do not treat "site" module as standard library module

* The way it was detected was also very wrong

kayhayen committed:
Windows: Make sure link library and exports are deleted

* Otherwise it can happen that plugins cause things like that added to standalone dist.

kayhayen committed:
Windows: Use hash filenames for C code of non-encodable module names

* Also remove the @1/@2 case disambiguation and use the hashes instead.

kayhayen committed:
Optimization: Faster encoding of identifiers for C code

kayhayen committed:
Tests: Minor spelling cleanup

kayhayen committed:
Python3.14: We don't do hacl inlining for that version anymore.

kayhayen committed:
Watch: Handle corrupt XML files properly

kayhayen committed:
Optimization: Was not recognizing module variable usages inside outline functions

* That prevented recognition of the case for caching to be there

kayhayen committed:
Optimization: Avoid taking reference for trivial constant returns

* We need to take a more full decision of what's immortal for best results

kayhayen committed:
Fix, need to prevent locals optimization when inlining the call to a locals dict variable

* After some scalability work this has come up as a possible path to mis-optimization

kayhayen committed:
Standalone: Added support for newer "dask" package

kayhayen committed:
New hotfix release.

kayhayen committed:
UI: Minor cleanup for memory traces

kayhayen committed:
UI: Enhanced tracing for DLL detection on Win32

kayhayen committed:
Fix, caching external use paths was ineffective

* This caused very slow resolutions to short paths to be executed a lot

* Also added a fast path for looking up an entry point that doesn't
  check paths deeply immediately.

* This makes DLL resolution on Windows much faster in the cached case
  where it was dominating the time and making it slow.

kayhayen committed:
Standalone: Added support for "win32ctypes" package

* It might use the "cffi" backend, check for its presence.

* Moved dependency to the actual package and not the using one.

kayhayen committed:
Windows: Don't attempt to find MSVC redist twice

kayhayen committed:
Fix, need to ignore non-existent packages from "top_level.txt" for metadata

kayhayen committed:
Merge branch 'hotfix/2.8.6'

kayhayen committed:
Merge tag '2.8.6' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
Release: Need to use lower names for sources archives of PyPI uploads now

kayhayen committed:
Next release will be 4.0, bumping version accordingly.

* We skip the Python3 number space in order to avoid confusion
  between them. Already Nuitka 2.7 could be confused with Python
  2.7 without proper context.

kayhayen committed:
UI: Added options for compile time profiling

* Also renamed old runtime profiling option, that is however still broken.

kayhayen committed:
Optimization: Much faster Python passes

* Escape and unknown traces now have their own
  number spaces. This allows to do some tests
  for a trace without using the actual object.

* Narrow the scope of variables to the outline scope
  that uses them, so that they don't need to be dealt
  with in merging later code where they don't ever
  change anymore and are not used at all.

* When checking for unused variables, do not ask the
  trace collection to filter its traces, instead work
  of the ones attached in the variable already. This
  avoids a lot of searching work. Also use a method
  to decide if a trace constitutes usage rather than
  a long elif chain.

* For "PASS 1" of "telethon.tl.types" which has been one
  of the known trouble makers with many classes and type
  annotations all changes combined improve the compilation
  time by 800%.

kayhayen committed:
Scons: Also allow Python versions only partially supported for Nuitka

* This enables using python3.14 for scons, which is needed for latest Fedora package builds

kayhayen committed:
UI: Fix, "--no-progressbar" wasn't affecting Scons

kayhayen committed:
Quality: Fix, wasn't passing assume yes for downloads for the commit hook

kayhayen committed:
Reports: Including CPU instr and cycle counters in timing on native Linux

* With appropriate configuration on Linux this allows to get at very
  precise timing configuration so we can judge even small compile time
  improvements correctly. We then don't need many runs to average out
  noise from other effects.

* Don't use wall clock but process time for steps that are not doing
  IO like module optimization for more accurate values otherwise, it
  is however not very accurate still.

kayhayen committed:
Optimization: Also decide writing traces for parameter variables faster

kayhayen committed:
UI: Avoid exception in tqdm progress bar during process shutdown.

kayhayen committed:
Optimization: Faster variable trace maintenance

* We now trace variables in trace collection as a dictionary
  per variable with a dictionary of the versions, this is
  closer to out frequent usage per variable.

* That makes it a lot easier to update variables after the
  tracing is finished to know their users and writers.

* Requires a lot less work, but also makes work less memory
  local such that the performance gain is relatively small
  despite less work being done.

* Also avoids that a set for the users is to be maintained.

kayhayen committed:
Optimization: Avoid local imports during module searches

* These are slightly slower and pollute the profiling opt the optimization phase

kayhayen committed:
Cleanups

* Avoid imports of whole modules and instead import only select names
  from them.

* For the nuitka.Options module we split off the states into a separate
  "nuitka.States" module that provides an object that is also faster
  to access that module attributes. This is a micro optimization that
  allows their use to be faster.

* Renamed XML reading/creating functions for clearer names when used
  with out the module name, same for source code references objects.

kayhayen committed:
Windows Store Python: Detect flavor properly and only reject accelerated mode

* Because the Python DLL cannot be used from the outside,
  it cannot be made to work without copying the Python
  DLL which is too ugly and an irrelevant configuration.

kayhayen committed:
Reports: Added interface for plugins to report information

Abdur-rahmaanJ committed:
Fix link to issue template in CONTRIBUTING.md

kayhayen committed:
Optimization: Avoid local imports because of plugins usage

* This is not all that is now possible, but a good step ahead.

* We put the callables to a new module and assign the "Plugins" class into it, making this very easy.

kayhayen committed:
Onefile: Use threads during compression

* This makes it much faster.

* Need to avoid using "tell" on them, but we can just simply
  check for sizes when it's closed instead, and in one instance
  we were doing that anyway already and just used it for the
  reporting.

kayhayen committed:
fixup! Fix link to issue template in CONTRIBUTING.md

kayhayen committed:
Optimization: Remove unused second argument for variable access collection

* We haven't made a different between read and write
  for while already.

* This gives a very minor performance improvement
  and is more of a cleanup solving a TODO.

kayhayen committed:
Optimization: Faster escaping of variables already in unknown state

* This mostly affects module variables which once they
  become unknown may frequently escape again.

* The number based check avoids the work for many cases.

* For the very trusted trace check we need to make that
  fast to test as well for best results.

kayhayen committed:
Optimization: Keep track of variables that can even escape

* Temporary and locals dict variables cannot escape yet were
  using a lot of time in virtual method calls.

* This avoids that time entirely making this step even faster.

* We do not take the effort to dynamically clean it up, so
  variables could be added there that then don't exist, but
  it's still faster this way as those are mostly rare cases.

kayhayen committed:
Optimization: Faster code generation

* This doesn't do indentation anymore unless asked by
  new option or in debug mode, saving 10% of the time
  use for generation of code in a larger module.

kayhayen committed:
Optimization: Remove assertions for frequently ran code

kayhayen committed:
Minor spelling cleanups

kayhayen committed:
Optimization: Avoid final extra micro pass

* Instead we relegate this to debug mode and check that
  it then doesn't do anything. For large modules this
  extra micro pass took up substantial time.

kayhayen committed:
Optimization: Faster assertion checks for value traces

* Since this is ran very frequently, make sure the test is
   as fast as possible.

kayhayen committed:
Optimization: Make sure modulo for unknown init values is true too

* Also shifted the responsibility for setting it to active
  to the function that sets it active.

* This improves performance even further since module
  variables in modules benefit from the fast code paths
  for not escaping them more often.

kayhayen committed:
Reports: Also including timing information for code generation

* With this per module value we can now see improvements
  there and problem makers.

kayhayen committed:
Minor spelling cleanup

kayhayen committed:
Minor cleanups

kayhayen committed:
UI: Remove warning for OrderedSet

* We no longer depend on a fast "OrderedSet" implementation for
  value tracing.

kayhayen committed:
Optimization: Faster variable tracing again

* Re-use start trace objects from previous micro passes,
  allocating all over is wasteful and harms cache locality,
  this gives a minor performance boost to avoid that.

* Specialize start traces to not have and use a "previous"
  trace, making some methods faster since they don't have
  to check, and start traces use up less memory as a result
  as well.

* Avoid parent class "__init__" calls in value traces, instead
  duplicate implementations. That's ugly but only faster.

* For some methods they now can be static methods with start
  traces, making their calls slightly faster, too.

* Use proper abstract methods instead of assertions against
  being called, that was old code.

kayhayen committed:
Fix, avoid setting an assign trace for removed assignment of mutable values

* Loop optimization was seeing this and then saw the
  change to uninit trace only one iteration later
  than possible. This could potentially cause a need
  for more micro passes than necessary.

kayhayen committed:
UI: Also report what plugin did cause the forbidden implicit import

kayhayen committed:
Standalone: Added support for "dataparser" package

* There was a missing data file.

kayhayen committed:
Optimization: Don't include tzdata on non-Windows

kayhayen committed:
fixup! Fix, need to ignore non-existent packages from "top_level.txt" for metadata

kayhayen committed:
macOS: Also remove extended attributes from data files.

* Actually just remove them from all files, they are not needed
  and this will be faster than calling it per file.

kayhayen committed:
Scons: Added newer inline copy with Scons that supports Visual Studio 2026

* Migrated and expanded our patches to avoid unused tools, deleted
  more than before.

* Remove annotations which is what we do now for inline copies, since
  they only waste space.

kayhayen committed:
Cleanup, remove annotations from inline copies

* They only waste memory, and it's a oneliner to do that with
  the "3to2" tool.

kayhayen committed:
Scons: Avoid using the "Python" tool here too

kayhayen committed:
macOS: Proper error exit for attempting to compile for arm64 on x86_64 python

kayhayen committed:
Reports: Don't crash when reporting before source directory is created

kayhayen committed:
Python2: Fix, need to import this exception

* Was crashing in code supposed to handle non-existent report files

kayhayen committed:
New hotfix release.

kayhayen committed:
Module: Fix, "pkgutil.iter_modules" wasn't working when loading into a namespace

kayhayen committed:
Merge branch 'hotfix/2.8.7'

kayhayen committed:
Merge tag '2.8.7' into develop

kayhayen committed:
New pre-release.

kayhayen committed:
fixup! Scons: Added newer inline copy with Scons that supports Visual Studio 2026

kayhayen committed:
New hotfix release.

kayhayen committed:
Merge branch 'hotfix/2.8.8'

kayhayen committed:
Merge tag '2.8.8' into develop

kayhayen committed:
fixup! Optimization: Much faster Python passes

kayhayen committed:
Python3.13: Fix, need to make sure we also export "PyInterpreter_*" API

kayhayen committed:
Debugging: Output what module had extra changes during debug extra micro pass

ZeroIntensity committed:
Python3.12+: Implement function type syntax

* This adds support for function generics.

---------

Co-authored-by: Kay Hayen <kay.hayen@gmail.com>

kayhayen committed:
fixup! Python3.12+: Implement function type syntax

ZeroIntensity committed:
Python3.14: Added groundwork for deferred evaluation of function annotations

* The Python side of things is delayed, but this
  is already good, and will stop us from using
  the original "annotate" of uncompiled functions
  accidentally.

ZeroIntensity committed:
Python3.14: Added support for uncompiled generator integration

* This is actually a crucial piece for correctness with
  asyncio and couroutines as well as asyncgen to properly
  integrate.

ZeroIntensity committed:
Python3.14: Follow change in order of evaluation of "__exit__" and "__enter__"

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Nuitka/Nuitka?pullRequest=3619) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you all sign our [Contributor License Agreement](https://cla-assistant.io/Nuitka/Nuitka?pullRequest=3619) before we can accept your contribution.<br/>**0** out of **3** committers have signed the CLA.<br/><br/>:x: kayhayen<br/>:x: ZeroIntensity<br/>:x: KRRT7<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Nuitka/Nuitka?pullRequest=3619) it.</sub>
